File: docs/website/integrations/locks.ts
// locks — Astro integration that tracks `<Lock id="...">` regions in MDX/MD.
//
// Scans `docs/website/src/content/docs/**/*.{md,mdx}`, hashes the inner
// content of every <Lock> block (SHA-256), and tracks per-block version
// history in `<repo>/.claude/locks/locks.json`. The store is gitignored
// and chmod 600'd on POSIX systems — convention-level "agents shouldn't
// poke at this directly," not real enforcement.
//
// Runs in two places:
//   * `astro:server:setup` — initial scan, then watches the docs tree
//     via the dev server's Vite watcher and re-scans only the changed
//     file on save. This is the loop the user wants — text edits don't
//     touch the Rust pipeline.
//   * `astro:build:start` — full scan on production builds. Parse
//     errors fail the build (unpaired/nested `<Lock>`, missing `id`).
//
// Companion CLI lives in Rust: `cargo run --release -p fce --example
// locks -- status | history | diff`. It reads the same JSON; the hash
// scheme is SHA-256 hex on both sides so values round-trip.

import type { AstroIntegration } from "astro";
import { promises as fs } from "fs";
import path from "path";
import { createHash } from "crypto";
import { fileURLToPath } from "url";

/**
 * Astro and Vite disagree on the type of the project root: Astro's
 * `config.root` is a `URL`, Vite's `server.config.root` is an absolute
 * path string. Normalise to a string.
 */
function rootToPath(root: URL | string): string {
  if (typeof root === "string") {
    return root.startsWith("file://") ? fileURLToPath(root) : root;
  }
  return fileURLToPath(root);
}

interface ParsedBlock {
  id: string;
  description?: string;
  comments?: string;
  content: string;
}

interface Snapshot {
  version: number;
  hash: string;
  content: string;
  saved_at: number;
  description?: string;
  comments?: string;
}

interface Block {
  post_id: string;
  lock_id: string;
  current_version: number;
  current_hash: string;
  current_content: string;
  description?: string;
  comments?: string;
  updated_at: number;
  history: Snapshot[];
}

interface Store {
  blocks: Block[];
}

const DOCS_REL = "src/content/docs";
const STORE_REL = ".claude/locks/locks.json";

function hashContent(s: string): string {
  return createHash("sha256").update(s, "utf8").digest("hex");
}

function parseAttr(attrs: string, name: string): string | undefined {
  const re = new RegExp(`(^|\\s)${name}="([^"]*)"`);
  const m = attrs.match(re);
  return m ? m[2] : undefined;
}

function parseLocks(text: string): { blocks: ParsedBlock[]; errors: string[] } {
  const blocks: ParsedBlock[] = [];
  const errors: string[] = [];
  let i = 0;
  const len = text.length;

  while (i < len) {
    const open = text.indexOf("<Lock", i);
    if (open === -1) break;

    // Confirm `<Lock` is followed by whitespace, `>`, or `/` so we
    // don't match `<Lockable`.
    const after = open + 5;
    if (after >= len) {
      errors.push(`file ends mid-<Lock at offset ${open}`);
      break;
    }
    const next = text[after];
    if (next !== " " && next !== "\t" && next !== "\n" && next !== "\r" && next !== ">" && next !== "/") {
      i = open + 1;
      continue;
    }

    const closeOpen = text.indexOf(">", after);
    if (closeOpen === -1) {
      errors.push(`unclosed <Lock tag near offset ${open}`);
      break;
    }
    if (closeOpen > 0 && text[closeOpen - 1] === "/") {
      errors.push(`<Lock /> is self-closing at offset ${open} — Lock must wrap content`);
      i = closeOpen + 1;
      continue;
    }

    const attrs = text.slice(after, closeOpen);
    const contentStart = closeOpen + 1;

    const closeStart = text.indexOf("</Lock>", contentStart);
    if (closeStart === -1) {
      errors.push(`<Lock> opened at offset ${open} never closed`);
      break;
    }

    const nestedOpen = text.indexOf("<Lock", contentStart);
    if (nestedOpen !== -1 && nestedOpen < closeStart) {
      errors.push(`nested Lock blocks not supported (outer at ${open}, inner at ${nestedOpen})`);
      i = closeStart + "</Lock>".length;
      continue;
    }

    const content = text.slice(contentStart, closeStart);
    const id = parseAttr(attrs, "id");
    if (!id) {
      errors.push(`<Lock> at offset ${open} missing required \`id\` attribute`);
      i = closeStart + "</Lock>".length;
      continue;
    }

    blocks.push({
      id,
      description: parseAttr(attrs, "description"),
      comments: parseAttr(attrs, "comments"),
      content,
    });

    i = closeStart + "</Lock>".length;
  }

  return { blocks, errors };
}

async function readStore(p: string): Promise<Store> {
  try {
    const s = await fs.readFile(p, "utf-8");
    const parsed = JSON.parse(s);
    if (parsed && Array.isArray(parsed.blocks)) return parsed as Store;
    return { blocks: [] };
  } catch {
    return { blocks: [] };
  }
}

async function writeStore(p: string, store: Store): Promise<void> {
  await fs.mkdir(path.dirname(p), { recursive: true });
  await fs.writeFile(p, JSON.stringify(store, null, 2));
  if (process.platform !== "win32") {
    try {
      await fs.chmod(p, 0o600);
    } catch {
      // best effort — chmod failure shouldn't break the dev server
    }
  }
}

async function walkDocs(dir: string): Promise<string[]> {
  const out: string[] = [];
  const stat = await fs.stat(dir).catch(() => null);
  if (!stat || !stat.isDirectory()) return out;
  const entries = await fs.readdir(dir, { withFileTypes: true });
  for (const e of entries) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) {
      out.push(...(await walkDocs(p)));
    } else if (e.isFile() && (e.name.endsWith(".mdx") || e.name.endsWith(".md"))) {
      out.push(p);
    }
  }
  return out;
}

function ingest(store: Store, postId: string, parsed: ParsedBlock, now: number): boolean {
  const hash = hashContent(parsed.content);
  const existing = store.blocks.find((b) => b.post_id === postId && b.lock_id === parsed.id);
  if (!existing) {
    store.blocks.push({
      post_id: postId,
      lock_id: parsed.id,
      current_version: 1,
      current_hash: hash,
      current_content: parsed.content,
      description: parsed.description,
      comments: parsed.comments,
      updated_at: now,
      history: [
        {
          version: 1,
          hash,
          content: parsed.content,
          saved_at: now,
          description: parsed.description,
          comments: parsed.comments,
        },
      ],
    });
    return true;
  }
  if (existing.current_hash === hash) {
    let changed = false;
    if (existing.description !== parsed.description) {
      existing.description = parsed.description;
      changed = true;
    }
    if (existing.comments !== parsed.comments) {
      existing.comments = parsed.comments;
      changed = true;
    }
    return changed;
  }
  const next = existing.current_version + 1;
  existing.current_version = next;
  existing.current_hash = hash;
  existing.current_content = parsed.content;
  existing.description = parsed.description;
  existing.comments = parsed.comments;
  existing.updated_at = now;
  existing.history.push({
    version: next,
    hash,
    content: parsed.content,
    saved_at: now,
    description: parsed.description,
    comments: parsed.comments,
  });
  return true;
}

async function scanFile(
  repoRoot: string,
  filePath: string,
  store: Store,
  now: number,
  errors: string[],
): Promise<boolean> {
  let text: string;
  try {
    text = await fs.readFile(filePath, "utf-8");
  } catch (e) {
    errors.push(`read ${filePath}: ${e instanceof Error ? e.message : String(e)}`);
    return false;
  }
  const { blocks, errors: parseErrors } = parseLocks(text);
  const rel = path.relative(repoRoot, filePath).split(path.sep).join("/");
  for (const e of parseErrors) errors.push(`${rel}: ${e}`);
  let changed = false;
  for (const b of blocks) {
    if (ingest(store, rel, b, now)) changed = true;
  }
  return changed;
}

interface RunPaths {
  projectRoot: string;
  repoRoot: string;
  docsDir: string;
  storePath: string;
}

function pathsFromProjectRoot(projectRoot: string): RunPaths {
  const repoRoot = path.resolve(projectRoot, "..", "..");
  return {
    projectRoot,
    repoRoot,
    docsDir: path.join(projectRoot, DOCS_REL),
    storePath: path.join(repoRoot, STORE_REL),
  };
}

export default function locks(): AstroIntegration {
  return {
    name: "locks",
    hooks: {
      "astro:server:setup": async ({ server, logger }) => {
        const projectRoot = rootToPath(server.config.root);
        const { repoRoot, docsDir, storePath } = pathsFromProjectRoot(projectRoot);

        const fullScan = async (): Promise<void> => {
          const now = Math.floor(Date.now() / 1000);
          const store = await readStore(storePath);
          const errors: string[] = [];
          const files = await walkDocs(docsDir);
          for (const f of files) {
            await scanFile(repoRoot, f, store, now, errors);
          }
          for (const e of errors) logger.warn(e);
          await writeStore(storePath, store);
        };

        await fullScan();
        logger.info(`scanned ${docsDir.replace(repoRoot + path.sep, "")} -> ${STORE_REL}`);

        const onChange = async (filePath: string): Promise<void> => {
          if (!filePath.endsWith(".mdx") && !filePath.endsWith(".md")) return;
          if (!filePath.startsWith(docsDir)) return;
          const now = Math.floor(Date.now() / 1000);
          const store = await readStore(storePath);
          const errors: string[] = [];
          await scanFile(repoRoot, filePath, store, now, errors);
          for (const e of errors) logger.warn(e);
          await writeStore(storePath, store);
        };

        server.watcher.on("change", onChange);
        server.watcher.on("add", onChange);
      },

      "astro:build:start": async ({ logger }) => {
        const projectRoot = process.cwd();
        const { repoRoot, docsDir, storePath } = pathsFromProjectRoot(projectRoot);

        const now = Math.floor(Date.now() / 1000);
        const store = await readStore(storePath);
        const errors: string[] = [];
        const files = await walkDocs(docsDir);
        for (const f of files) {
          await scanFile(repoRoot, f, store, now, errors);
        }
        if (errors.length > 0) {
          for (const e of errors) logger.error(e);
          throw new Error(`locks: ${errors.length} parse error(s) — see above`);
        }
        await writeStore(storePath, store);
        logger.info(`scanned ${files.length} file(s) -> ${STORE_REL}`);
      },
    },
  };
}

File: scripts/locks.rs
mod locks {
    //! Build-time scanner for `<Lock id="...">` regions in MDX/MD content.
    //!
    //! Walks `docs/website/src/content/docs/**/*.{mdx,md}`, finds every
    //! `<Lock id="..." description="..." comments="...">...</Lock>` block,
    //! validates that every open tag has a matching close (and that locks
    //! don't nest), hashes the inner content with SipHash-1-3 (std), and
    //! tracks per-block version history in `.claude/locks/locks.json`.
    //! The store file is chmod 600'd after every write — a convention-level
    //! "agents shouldn't poke at this directly" signal more than a security
    //! boundary (any process running as the user can override).
    //!
    //! Failures (unpaired tag, nested lock, missing `id` attribute) print a
    //! diagnostic and exit the build. Hash drift does NOT fail the build —
    //! it just appends a new history entry. Restoration / strict drift
    //! enforcement is a follow-up tier.
    //!
    //! Companion CLI: `cargo run --release -p fce --example locks -- ...`
    //! reads the store and surfaces history + diffs.
    //!
    //! Re-run trigger: only the docs content directory; this scan is cheap
    //! (string scan + tiny JSON write) so re-running on every doc edit is
    //! fine.

    use std::collections::hash_map::DefaultHasher;
    use std::fs;
    use std::hash::{Hash, Hasher};
    use std::path::{Path, PathBuf};
    use std::time::{SystemTime, UNIX_EPOCH};

    use serde::{Deserialize, Serialize};

    const DOCS_DIR: &str = "docs/website/src/content/docs";
    const STORE_REL: &str = ".claude/locks/locks.json";

    #[derive(Default, Serialize, Deserialize)]
    struct Store {
        blocks: Vec<Block>,
    }

    #[derive(Serialize, Deserialize, Clone)]
    struct Block {
        post_id: String,
        lock_id: String,
        current_version: u64,
        current_hash: String,
        current_content: String,
        description: Option<String>,
        comments: Option<String>,
        updated_at: u64, // unix seconds
        history: Vec<Snapshot>,
    }

    #[derive(Serialize, Deserialize, Clone)]
    struct Snapshot {
        version: u64,
        hash: String,
        content: String,
        saved_at: u64, // unix seconds
        description: Option<String>,
        comments: Option<String>,
    }

    pub fn run() {
        let workspace = super::meta::workspace_root();
        let docs_dir = workspace.join(DOCS_DIR);
        if !docs_dir.exists() {
            return; // no docs in this checkout — nothing to scan
        }

        let store_path = workspace.join(STORE_REL);
        if let Some(parent) = store_path.parent() {
            let _ = fs::create_dir_all(parent);
        }

        let mut store = read_store(&store_path);

        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);

        // Walk and parse every MDX/MD under DOCS_DIR.
        let mut errors: Vec<String> = Vec::new();
        let mut seen: std::collections::HashSet<(String, String)> = std::collections::HashSet::new();

        for path in walk_docs(&docs_dir) {
            let rel = path
                .strip_prefix(&workspace)
                .unwrap_or(&path)
                .to_string_lossy()
                .to_string();
            let text = match fs::read_to_string(&path) {
                Ok(t) => t,
                Err(e) => {
                    errors.push(format!("read {}: {}", rel, e));
                    continue;
                }
            };
            let blocks = match parse_locks(&text) {
                Ok(b) => b,
                Err(e) => {
                    errors.push(format!("{}: {}", rel, e));
                    continue;
                }
            };
            for b in blocks {
                seen.insert((rel.clone(), b.id.clone()));
                ingest(&mut store, &rel, &b, now);
            }
        }

        if !errors.is_empty() {
            for e in &errors {
                eprintln!("locks: {}", e);
            }
            // Fail the build: unpaired or malformed Lock blocks are bugs.
            panic!("scripts/locks.rs: {} error(s) — see above", errors.len());
        }

        write_store(&store_path, &store);

        // chmod 600 on unix — convention-level deterrent.
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            if let Ok(meta) = fs::metadata(&store_path) {
                let mut perms = meta.permissions();
                perms.set_mode(0o600);
                let _ = fs::set_permissions(&store_path, perms);
            }
        }

        println!("cargo::rerun-if-changed={}", DOCS_DIR);
    }

    fn ingest(store: &mut Store, post_id: &str, parsed: &ParsedBlock, now: u64) {
        let hash = hash_content(&parsed.content);
        let pos = store
            .blocks
            .iter()
            .position(|b| b.post_id == post_id && b.lock_id == parsed.id);
        match pos {
            None => {
                let snap = Snapshot {
                    version: 1,
                    hash: hash.clone(),
                    content: parsed.content.clone(),
                    saved_at: now,
                    description: parsed.description.clone(),
                    comments: parsed.comments.clone(),
                };
                store.blocks.push(Block {
                    post_id: post_id.to_string(),
                    lock_id: parsed.id.clone(),
                    current_version: 1,
                    current_hash: hash,
                    current_content: parsed.content.clone(),
                    description: parsed.description.clone(),
                    comments: parsed.comments.clone(),
                    updated_at: now,
                    history: vec![snap],
                });
            }
            Some(i) => {
                let block = &mut store.blocks[i];
                if block.current_hash == hash {
                    // Content unchanged. Refresh metadata if attrs changed.
                    if block.description != parsed.description {
                        block.description = parsed.description.clone();
                    }
                    if block.comments != parsed.comments {
                        block.comments = parsed.comments.clone();
                    }
                    return;
                }
                let next_version = block.current_version + 1;
                let snap = Snapshot {
                    version: next_version,
                    hash: hash.clone(),
                    content: parsed.content.clone(),
                    saved_at: now,
                    description: parsed.description.clone(),
                    comments: parsed.comments.clone(),
                };
                block.current_version = next_version;
                block.current_hash = hash;
                block.current_content = parsed.content.clone();
                block.description = parsed.description.clone();
                block.comments = parsed.comments.clone();
                block.updated_at = now;
                block.history.push(snap);
            }
        }
    }

    fn read_store(path: &Path) -> Store {
        match fs::read_to_string(path) {
            Ok(s) => serde_json::from_str(&s).unwrap_or_default(),
            Err(_) => Store::default(),
        }
    }

    fn write_store(path: &Path, store: &Store) {
        match serde_json::to_string_pretty(store) {
            Ok(s) => {
                if let Err(e) = fs::write(path, s) {
                    eprintln!("locks: write {} failed: {}", path.display(), e);
                }
            }
            Err(e) => eprintln!("locks: serialize failed: {}", e),
        }
    }

    fn walk_docs(dir: &Path) -> Vec<PathBuf> {
        let mut out = Vec::new();
        if let Ok(entries) = fs::read_dir(dir) {
            for entry in entries.flatten() {
                let p = entry.path();
                if p.is_dir() {
                    out.extend(walk_docs(&p));
                } else if let Some(ext) = p.extension() {
                    if ext == "mdx" || ext == "md" {
                        out.push(p);
                    }
                }
            }
        }
        out
    }

    fn hash_content(content: &str) -> String {
        let mut h = DefaultHasher::new();
        content.hash(&mut h);
        format!("{:016x}", h.finish())
    }

    struct ParsedBlock {
        id: String,
        description: Option<String>,
        comments: Option<String>,
        content: String,
    }

    fn parse_locks(text: &str) -> Result<Vec<ParsedBlock>, String> {
        let bytes = text.as_bytes();
        let len = bytes.len();
        let mut out = Vec::new();
        let mut i = 0;

        while i < len {
            let rel = match find_from(bytes, i, b"<Lock") {
                Some(n) => n,
                None => break,
            };
            // Confirm `<Lock` is followed by whitespace, `>`, or `/` — not
            // another identifier character (so we don't match `<Lockable`).
            let after = rel + 5;
            if after >= len {
                return Err("file ends mid-Lock tag".into());
            }
            let next = bytes[after];
            if !(next.is_ascii_whitespace() || next == b'>' || next == b'/') {
                i = rel + 1;
                continue;
            }

            // Find end of the open tag.
            let close_open = find_from(bytes, after, b">")
                .ok_or_else(|| format!("unclosed `<Lock` near byte {}", rel))?;

            // Self-closing `<Lock ... />` — locks must wrap content; reject.
            if close_open > rel && bytes[close_open - 1] == b'/' {
                return Err(format!(
                    "`<Lock />` is self-closing at byte {} — Lock must wrap content",
                    rel
                ));
            }

            let attrs = std::str::from_utf8(&bytes[after..close_open])
                .map_err(|e| format!("non-utf8 in attrs near byte {}: {}", rel, e))?;
            let content_start = close_open + 1;

            // Find the matching `</Lock>`. Disallow nesting: any `<Lock`
            // before the close fails the parse.
            let close_rel = find_from(bytes, content_start, b"</Lock>")
                .ok_or_else(|| format!("`<Lock>` opened at byte {} never closed", rel))?;
            if let Some(nested) = find_from(bytes, content_start, b"<Lock") {
                if nested < close_rel {
                    return Err(format!(
                        "nested Lock blocks not supported (outer opens at byte {}, inner at {})",
                        rel, nested
                    ));
                }
            }

            let content = std::str::from_utf8(&bytes[content_start..close_rel])
                .map_err(|e| format!("non-utf8 content near byte {}: {}", rel, e))?;

            let id = parse_attr(attrs, "id")
                .ok_or_else(|| format!("`<Lock>` near byte {} missing required `id` attribute", rel))?;
            let description = parse_attr(attrs, "description");
            let comments = parse_attr(attrs, "comments");

            out.push(ParsedBlock {
                id,
                description,
                comments,
                content: content.to_string(),
            });

            i = close_rel + b"</Lock>".len();
        }

        Ok(out)
    }

    /// Find `needle` in `haystack[from..]` and return the absolute index, or None.
    fn find_from(haystack: &[u8], from: usize, needle: &[u8]) -> Option<usize> {
        if from >= haystack.len() {
            return None;
        }
        haystack[from..]
            .windows(needle.len())
            .position(|w| w == needle)
            .map(|n| from + n)
    }

    /// Parse a double-quoted attribute value out of an attribute string.
    /// Accepts `name="..."`. Does NOT support `name={...}` JSX-expression
    /// syntax for the first iteration — the build script will surface a
    /// clear error if `id` is missing.
    fn parse_attr(attrs: &str, name: &str) -> Option<String> {
        let target = format!("{}=", name);
        let mut start = 0;
        while let Some(rel) = attrs[start..].find(&target) {
            let pos = start + rel;
            // Make sure the character before is a word boundary (start of
            // the attrs or whitespace) so `id` doesn't match e.g. `pid`.
            let before_ok = pos == 0
                || attrs.as_bytes()[pos - 1].is_ascii_whitespace();
            if !before_ok {
                start = pos + target.len();
                continue;
            }
            let after = &attrs[pos + target.len()..];
            let after = after.trim_start();
            if let Some(rest) = after.strip_prefix('"') {
                if let Some(end) = rest.find('"') {
                    return Some(rest[..end].to_string());
                }
            }
            // `name={...}` is unsupported — bail this match, try next.
            start = pos + target.len();
        }
        None
    }
}
